### PR TITLE
refactor: ScheduleQueryButton into functional component

### PR DESCRIPTION
**Pull request by AAfghahi**
### SUMMARY
Changed the ScheduleQueryButton into a TypeScript function using react hooks. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Screen Shot 2021-03-03 at 1 40 38 PM

### CODE REVIEW

File: docker/pythonpath_dev/superset_config.py [hunk 1]
```diff
@@ -78,7 +78,95 @@ class CeleryConfig(object):
...
+        # that uses the query id and the output table as its name
+        'linkback': (
+            'https://airflow.example.com/admin/airflow/tree?'
+            'dag_id=query_${id}_${extra_json.schedule_info.output_table}'
+        ),
+    },
+}
+WTF_CSRF_ENABLED = False
+ENABLE_SCHEDULED_EMAIL_REPORTS = True
+ENABLE_ALERTS = True
```
Comment (eschutho) [inline docker/pythonpath_dev/superset_config.py:169]: you'll want to move this into you superset_config in your root folder so that it's not committed.

Comment (AAfghahi) [reply docker/pythonpath_dev/superset_config.py:169]: is that a folder that I have to create? I don't see it.

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  dbId,
+  onSchedule = () => {},
+  scheduleQueryWarning = null,
+  tooltip = '',
+}) => {
+  const [description, setDescription] = useState('');
+  const [label, setLabel] = useState(defaultLabel);
+  const [showSchedule, setShowSchedule] = useState(false);
+  let saveModal: any;
+  // this is for the ref that is created in the modal trigger. the modal is created in order to use the .close() function on it.
```
Comment (hughhhh) [inline superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:123]: is this comment still valid?

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+import Button from 'src/components/Button';
+
+const validators = {
+  greater: (a: number, b: number) => a > b,
+  greater_equal: (a: number, b: number) => a >= b,
+  less: (a: number, b: number) => a < b,
+  less_equal: (a: number, b: number) => a <= b,
+};
+
+function getJSONSchema() {
```
Comment (hughhhh) [inline superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:36]:
```suggestion
const getJSONSchema = () => {
```

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+          };
+        }
+      },
+    );
+    return jsonSchema;
+  }
+  return {};
+}
+
+function getUISchema() {
```
Comment (hughhhh) [inline superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:55]:
```
const getUISchema = () => {
```

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  return window.featureFlags.SCHEDULED_QUERIES?.UISCHEMA;
+}
+
+function getValidationRules() {
```
Comment (hughhhh) [inline superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:59]:
```suggestion
const getValidationRules = () => {
```

**Review by hughhhh — APPROVED** [review]

File: docker/pythonpath_dev/superset_config.py [hunk 6]
```diff
@@ -78,7 +78,6 @@ class CeleryConfig(object):
 
 CELERY_CONFIG = CeleryConfig
 SQLLAB_CTAS_NO_LIMIT = True
-
```
Comment (eschutho) [inline docker/pythonpath_dev/superset_config.py:81]: just for code cleanliness, I would revert this change.

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+  scheduleQueryWarning: string | null;
+  disabled: boolean;
+  tooltip: string;
+}
+
+const StyledRow = styled(Row)`
+  padding-bottom: ${({ theme }) => theme.gridUnit * 2}px;
+`;
+
+const ButtonComponent = styled.div`
```
Comment (eschutho) [inline superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:91]: where did these styles come from? Are they new or copied?

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  }
+`;
+
+const ScheduleQueryButton: FunctionComponent<ScheduleQueryButtonProps> = ({
+  defaultLabel = t('Undefined'),
+  sql,
+  schema,
+  dbId,
+  onSchedule = () => {},
+  scheduleQueryWarning = null,
```
Comment (eschutho) [inline superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:113]: per the props above this will never be undefined so you're not going to hit this default. If you feel you need this default because there's not enough type checking going on with the other files, then you can set the typescript prop as optional, but not sure if you need it. You may have to look through the code to see if it's there intentionally.

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+  tooltip = '',
```
Comment (eschutho) [inline superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:114]: same thing here

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx (hunk 7 above)
Comment (AAfghahi) [reply superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:91]: I copied them from what the styled divs looked like. It is because I switched this from being a div to being a button component because there were TypeScript issues with the div component having an onClick and a disabled option.

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+}) => {
+  const [description, setDescription] = useState('');
+  const [label, setLabel] = useState(defaultLabel);
+  const [showSchedule, setShowSchedule] = useState(false);
+  let saveModal: any;
```
Comment (eschutho) [inline superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:119]: suggestion: can you make this a type of ref or undefined?

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+  const onScheduleSubmit = ({ formData }: Record<string, any>) => {
```
Comment (eschutho) [inline superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:121]: does this formData have a type? Can you do something like `({ formData }: {formData: HTMLFormData})`? That's just a guess on the type

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx (hunk 7 above)
Comment (eschutho) [reply superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:91]: You can also (for consistency) call this `StyledButtonComponent` and then instead of use a div to wrap the button, you can do styled(Button)` and then you don't have to wrap this around the button element, but just use this as the tag.
Example: https://github.com/apache/superset/blob/9a05d6afe537d9b2bc605f515590c8461a8e95e4/superset-frontend/src/explore/components/DataTableControl.tsx#L38
(Assume it's called StyledCopyButton though :))

File: superset-frontend/src/featureFlags.ts [hunk 12]
```diff
@@ -43,8 +43,40 @@ export enum FeatureFlag {
...
+  DISABLE_DATASET_SOURCE_EDIT?: boolean;
+  DISPLAY_MARKDOWN_HTML?: boolean;
+  ESCAPE_MARKDOWN_HTML?: boolean;
+  DASHBOARD_NATIVE_FILTERS?: boolean;
+  DASHBOARD_CROSS_FILTERS?: boolean;
+  DASHBOARD_NATIVE_FILTERS_SET?: boolean;
+  VERSIONED_EXPORT?: boolean;
+  GLOBAL_ASYNC_QUERIES?: boolean;
+  ENABLE_TEMPLATE_PROCESSING?: boolean;
+  ENABLE_EXPLORE_DRAG_AND_DROP?: boolean;
```
Comment (eschutho) [inline superset-frontend/src/featureFlags.ts:79]: You should be able to do this instead: 
```
export type FeatureFlagMap = {
  [key in Exclude<FeatureFlag, FeatureFlag.SCHEDULED_QUERIES>]: boolean;
} & {
  SCHEDULED_QUERIES: ScheduleQueriesProps;
};
```

Comment (AAfghahi) [reply superset-frontend/src/featureFlags.ts:79]: this looks way better, thank you so much!

**Review by eschutho — CHANGES_REQUESTED** [review]
few other suggestions..

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+          saveModal = ref;
+        }}
+        modalTitle={t('Schedule query')}
+        modalBody={renderModalBody()}
+        triggerNode={
+          <ButtonComponent>
+            <Button
+              onClick={() => setShowSchedule(!showSchedule)}
+              buttonSize="small"
+              tooltip={tooltip}
```
Comment (eschutho) [inline superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:197]: I think we lost `disabled` here.. also you can add `buttonStyle="link"` and it will get you most of the styles that you need.

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx (hunk 10 above)
Comment (AAfghahi) [reply superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:119]: Done! What's the reasoning behind it being undefined though?

File: superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx (hunk 11 above)
Comment (AAfghahi) [reply superset-frontend/src/SqlLab/components/ScheduleQueryButton.tsx:121]: is this a second deconstruction? {formData: HTMLFormData}

**Comment by eschutho** [comment]
@hughhhh will finish the review on this.. thanks @AAfghahi for everything!
